File: src/serialize/audit_log.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 The MCDF Project
#include "mcdf/serialize/audit_log.hpp"

namespace mcdf {

std::string audit_entry_to_line(const AuditEntry& entry) {
  return entry.timestamp + "\t" + entry.action + "\t" + entry.actor + "\t" +
         entry.prev_hash;
}

Result<std::vector<AuditEntry>> parse_audit_log(std::string_view text) {
  std::vector<AuditEntry> out;
  std::size_t start = 0;
  while (start < text.size()) {
    std::size_t nl = text.find('\n', start);
    const std::string_view line =
        (nl == std::string_view::npos) ? text.substr(start)
                                       : text.substr(start, nl - start);
    start = (nl == std::string_view::npos) ? text.size() : nl + 1;
    if (line.empty()) continue;

    std::string fields[4];
    std::size_t field = 0;
    std::size_t pos = 0;
    bool too_many = false;
    while (pos <= line.size()) {
      const std::size_t tab = line.find('\t', pos);
      const std::string_view part =
          (tab == std::string_view::npos) ? line.substr(pos)
                                          : line.substr(pos, tab - pos);
      if (field >= 4) { too_many = true; break; }
      fields[field++] = std::string(part);
      if (tab == std::string_view::npos) break;
      pos = tab + 1;
    }
    if (too_many || field != 4) {
      return fail(ErrorCode::kParse, "malformed audit.log entry");
    }
    out.push_back({fields[0], fields[1], fields[2], fields[3]});
  }
  return out;
}

}  // namespace mcdf

File: apps/web/packages/mcdf-ts/src/serialize/audit-log.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2026 Sukesh Ashok Kumar and The MCDF Project
//
// `audit.log` line format, ported from `src/serialize/audit_log.cpp`.
//
// One tab-separated line per entry: timestamp, action, actor, prev_hash. The
// line is hashed verbatim to chain the next entry, so its exact bytes matter as
// much as any other part of the format.

import type { AuditEntry } from '../model/types.js';
import { ParseError } from './yaml.js';

export const AUDIT_LOG_PATH = 'audit.log';
export const AUDIT_CHECKPOINT_PATH = 'audit.checkpoint';

/** The prev_hash of the first entry: 64 zeros. */
export const AUDIT_GENESIS = '0'.repeat(64);

export function auditEntryToLine(entry: AuditEntry): string {
  return `${entry.timestamp}\t${entry.action}\t${entry.actor}\t${entry.prev_hash}`;
}

/**
 * Parses the whole log. A line with the wrong number of fields is a parse
 * error, not a skipped line: silently dropping an unreadable entry would break
 * the chain and report the break as if the file were shorter than it is.
 */
export function parseAuditLog(text: string): AuditEntry[] {
  const entries: AuditEntry[] = [];
  for (const line of text.split('\n')) {
    if (line === '') continue;
    const fields = line.split('\t');
    if (fields.length !== 4) {
      throw new ParseError('malformed audit.log entry');
    }
    entries.push({
      timestamp: fields[0] as string,
      action: fields[1] as string,
      actor: fields[2] as string,
      prev_hash: fields[3] as string,
    });
  }
  return entries;
}
